dask_cloudprovider/azure/utils.py

```diff
@@ -0,0 +1,178 @@
+import asyncio
+import datetime
+import logging
+
+import aiohttp
+from distributed.diagnostics.plugin import WorkerPlugin
+from tornado.ioloop import IOLoop, PeriodicCallback
+
+
+logger = logging.getLogger(__name__)
+
+
+class AzurePreemptibleWorkerPlugin(WorkerPlugin):
+    """A worker plugin for azure spot instances
+
+    This worker plugin will poll azure's metadata service for preemption notifications.
+    When a node is preempted, the plugin will attempt to shutdown gracefully all workers
+    on the node.
+
+    For more details on azure spot instances see:
+    https://docs.microsoft.com/en-us/azure/virtual-machines/linux/scheduled-events
+
+    Parameters
+    ----------
+    poll_interval_s: int (optional)
+        The rate at which the plugin will poll the metadata service in seconds.
+
+        Defaults to ``1``
+
+    metadata_url: str (optional)
+        The url of the metadata service to poll.
+
+        Defaults to "http://169.254.169.254/metadata/scheduledevents?api-version=2019-08-01"
+
+    termination_events: List[str] (optional)
+        The type of events that will trigger the gracefull shutdown
+
+        Defaults to ``['Preempt', 'Terminate']``
+
+    termination_offset_minutes: int (optional)
+        Extra offset to apply to the premption date. This may be negative, to start
+        the gracefull shutdown before the ``NotBefore`` date. It can also be positive, to
+        start the shutdown after the ``NotBefore`` date, but this is at your own risk.
+
+        Defaults to ``0``
+
+    Examples
+    --------
+
+    Let's say you have cluster and a client instance. For example using ``KubeCluster``
```

Comment:
Let's make this a full reference to span the two sets of docs. Will need to add dask-kubernetes to the intersphinx mapping.

https://github.com/dask/dask-cloudprovider/blob/dffd81d3e4ab6feb5378ffa0d9952f3c7da207f7/doc/source/conf.py#L194-L198

```suggestion
    Let's say you have cluster and a client instance. For example using :class:`dask_kubernetes.KubeCluster`
```